#!/bin/bash 

REFDIR=$1
TAG=$2
GFF=$3
FA=$4

## the following directory structure is implied: 

##$REFDIR/Assemblies 
##$REFDIR/bowtie2
##$REFDIR/RSEM
##$REFDIR/kallisto

## 1) check what sort of GTF file you're dealing with (if it has ^> in it)
## we assume there's always ncRNA - if that's empty, scripts should still work 
 
PP1=`grep -P "^>" $GTF`
PP2=`grep -P "##FASTA" $GTF`

if [[ $PP1 != "" && $PP2 != "" && $FA == "" ]]
then 
  echo "The GTF file provided was identified as generated by Prokka! Continuing.." 
  N_CDS=`grep -c -P "\tCDS\t" $GTF`
  N_NCR=`grep -c -P "\tmisc_RNA\t" $GTF`
  echo "Found $N_CDS protein-coding (CDS) and $N_NCR non-coding RNA (misc_RNA) features!"
  grep -P "\tCDS\t" $GTF > $REFDIR/$TAG.CDS.gff
  grep -P "\tmisc_RNA\t" $GTF | sed "s/misc_RNA/ncRNA/g" > $REFDIR/$TAG.ncRNA.gff 
  grep -v -P "^#" $GTF | grep -v "\t" $GTF > $REFDIR/$TAG.genome.fa
  echo "Files $TAG.genome.fa, $TAG.CDS.gff and $TAG.ncRNA.gff successfully generated in $REFDIR!"
elsif [[ $PP1 == "" && $PP2 == "" && $FA != "" ]]
then
  echo "Custom GTF/FA files provided! Continuing.."
  N_CDS=`grep -c -P "\tCDS\t" $GTF`
  N_NCR=`grep -c -P "\tmisc_RNA\t|\tncRNA\t" $GTF`
  echo "Found $N_CDS protein-coding (CDS) and $N_NCR non-coding RNA (misc_RNA/ncRNA) features!"
  grep -P "\tCDS\t" $GTF > $REFDIR/$TAG.CDS.gff
  grep -P "\tmisc_RNA\t|\tncRNA\t" $GTF | sed "s/misc_RNA/ncRNA/g" > $REFDIR/$TAG.ncRNA.gff 
  cp $FA $REFDIR/$TAG.genome.fa
  echo "Files $TAG.genome.fa, $TAG.CDS.gff and $TAG.ncRNA.gff successfully generated in $REFDIR!"
else
  echo "ERROR: Either provide a Prokka-style GTF file, or a custom GTF/FA pair!"
fi 

## 3) make 3-col file with all of the sub-fields 
cat $REFDIR/$TAG.CDS.gff $REFDIR/$TAG.ncRNA.gff | \
perl -ne '$n; $id; @t=split/\t/; $t[8]=~m/^ID=(.*?);.*/; $id=$1; if ($t[8]=~m/.*;Name=(.*?);.*/) {$n=$1} else {$n=$id}; print "$id\t$n\t$t[2]\n"' > $TAG.3col 

## 4) run prokka on genome, find all of the rRNA sequences 
source activate prokka 
prokka --cpus 16 --outdir $TAG.prokka --prefix $TAG.prokka --locustag $TAG $REFDIR/$FA
source deactivate

grep -P "\ttRNA\t" $TAG.prokka/$TAG.prokka.gff | awk '{print $1"\t"$4-1"\t"$5"\ttRNA."NR"\t.\t"$7}' > $TAG.tRNA.bed
grep -P "\trRNA\t" $TAG.prokka/$TAG.prokka.gff | awk '{print $1"\t"$4-1"\t"$5"\trRNA."NR"\t.\t"$7}' > $TAG.rRNA.bed
bedtools getfasta -name -s -fi $REFDIR/$TAG.genome.fa -bed $TAG.tRNA.bed -fo $TAG.tRNA.fa
bedtools getfasta -name -s -fi $REFDIR/$TAG.genome.fa -bed $TAG.rRNA.bed -fo $TAG.rRNA.fa
rm -rf $TAG.prokka

## 5) *chrom.size reference 

samtools faidx $REFDIR/$TAG.genome.fa
cut -f 1,2 $TAG.genome.fa.fai > $TAG.chrom.sizes 

## 6) bowtie2 reference 
bowtie2-build $REFDIR/$TAG.genome.fa $TAG
bowtie2-build $TAG.tRNA.fa $TAG.tRNA
bowtie2-build $TAG.rRNA.fa $TAG.rRNA 

## 8) rsem reference - skip this for now

## 9) copy to appropriate locations

