# multi-bacpipe
A pipeline for uniform multi-strain RNA-seq processing.
<img align="right" width="407" height="229" src="http://static.bnr.bg/sites/en/music/publishingimages/630/12-06-21-81065_2.jpg">

## Author
[Alexander Predeus](https://www.researchgate.net/profile/Alexander_Predeus), [Jay Hinton Laboratory](http://www.hintonlab.com/), [University of Liverpool](https://www.liverpool.ac.uk/)

(c) 2018, GPL v3 license

## Motivation
This is an upgrade to [bacpipe](https://github.com/apredeus/bacpipe) that is capable of performing uniform RNA-seq appropriate annotation of muliple strains and/or genomes, and then create master expression tables streamlined for a detailed strain comparison. 

When successfully applied, this should generate (for each strain):
* genomic bam files for read-resolution visualization and analysis;
* TDF files for visualization in IGV;
* scaled bigWig files for visualization in JBrowse (see [doi:10.1128/mBio.01442-14](http://mbio.asm.org/content/5/4/e01442-14.full) for description of scaling); 
* raw read and TPM expression tables - from [featureCounts](http://subread.sourceforge.net/), [rsem](https://deweylab.github.io/RSEM/), and [kallisto](https://pachterlab.github.io/kallisto/); 
* a single [MultiQC](http://multiqc.info/) quality control report.

This should also generate several master tables for all sorts of downstream analysis (clustering, PCA, differential expression, etc).

## Installation and requirements 
Clone the pipeline scripts into your home directory and add them to $PATH variable in bash: 

```bash
cd ~
git clone https://github.com/apredeus/multi-bacpipe
echo "export PATH=~/multi-bacpipe:~/multi-bacpipe/utils:\$PATH" >> ~/.bashrc
```

To install the requirements, use [Bioconda](https://bioconda.github.io/). Below are the programs that need to be installed. In our experience, it helps to have roary and prokka installed in their own virtual environments. Management of virtual environments in Bioconda is described [here](https://conda.io/docs/user-guide/tasks/manage-environments.html).

```bash 
conda install fastqc
conda install star
conda install samtools
conda install bedtools
conda install igvtools
conda install subread
conda create -n prokka -c conda-forge -c bioconda prokka
conda create -n roary  -c conda-forge -c bioconda roary
```
You also need to have Perl installed. Sorry. 

## Setting up the working directory
After you have chosen your working directory (with enough space and where you have write permissions), you need to create a sub-directory named **fastqs**. 
Place your archived RNA-seq *fastq* files here. Stick to the following naming rules: 

* Single-end sequencing: **Sample_ID.fastq.gz**; 
* Paired-end sequencing: **Sample_ID.R1.fastq.gz**, **Sample_ID.R2.fastq.gz**. 

If you have more than one file per sample (e.g. if it was sequenced twice), it's recommended to merge these files first.

## Configuration file
Multi-strain processing is relying onto single-strain processing principles, described in [Bacpipe README](https://github.com/apredeus/bacpipe). In order to process multiple strains, you would need to make a simple tab-separated configuration file. The file should contain one tab-separated sample ID - strain ID pair per line. It should also include all of the strains that you want to use as a reference: 

```
Sample_ID1	Ecoli_O157_H7
Sample_ID2	Ecoli_O157_H7
Sample_ID3	Ecoli_O104_H4
Sample_ID4	Ecoli_O104_H4
Reference	MG1655
Reference	DH5a
Reference	DE3
```

See details on reference preparation below. 

## Two-step reference preparation
Reference preparation includes two steps: 
* Prepare reference for all strains you are using in your RNA-seq using **prepare_bacpipe_reference.sh** (in the config example above, these are Ecoli_O157_H7 and Ecoli_O104_H4). See more details at [Bacpipe README](https://github.com/apredeus/bacpipe); 
* Prepare combined reference using select reference strains (in the config example above, MG1655, DH5a, and DE3) by running **prepare_multireference.sh**. 

`prepare_bacpipe_reference.sh <reference_dir> <tag> <name>.gff <name>.fa`

or 

`prepare_bacpipe_reference.sh <reference_dir> <tag> <name>.gff` 

for Prokka-style GFF file.

## Prophages and rRNA operons
In order to provide additional information, the strains you're using in your RNA-seq should be include two additional *bed* files:
* Putative prophage intervals - these can be generated by parsing the output of [PHASTER](http://phaster.ca/); 
* rRNA operon borders - I am working on correctly guessing the borders automatically, but for now, we're going to need a *bed* file. This is very important in order to accurately calculate the stats, especially the percentage of multi-mapping and un-assigned reads **after rRNA removal**.

## One-command RNA-seq processing
After all the references are successfully created, simply run 

`multi-bacpipe.sh  <reference_dir>  <config_file>  <CPUs>`

Bacpipe needs to be ran in a writeable directory with fastqs folder in it. 

Bacpipe:
* handles archived (.gz) and non-archived fastq files; 
* handles single-end and paired-end reads; 
* automatically detects strand-specificity of the experiment; 
* performs quantification according to the calculated parameters. 

The following steps are performed during the pipeline execution: 
* FastQC is ran on all of the fastq files; 
* STAR is used to align the fastq files to the rRNA and tRNA reference to accurately estimate rRNA/tRNA content; 
* Unsorted bam alignments are filtered using rRNA *bed* file, sorted, and indexed; 
* tdf files are prepared for visualization in IGV; 
* bigWig (bw) files are prepared for vizualization in majority of other genomic browsers; 
* featureCounts is ran on genomic bam to evaluate the strandedness of the experiment; 
* strandedness and basic alignment statistics are calculated; 
* featureCounts output is chosen based correct settings of strandedness; 
* appropriately formatted logs are generated; 
* multiqc is ran to summarize everything as a nicely formatted report. 
    
In the end you are expected to obtain a number of new directories: FastQC, bams, tdfs_and_bws, stats, strand, featureCounts, exp_tables. Each directory would contain the files generated by its namesake, as well as all appropriate logs. The executed commands with all of the versions and exact options are recorded in the master log. 
    
    
